src/SmartEnum/SmartEnum.cs

```diff
@@ -35,39 +35,29 @@ public abstract class SmartEnum<TEnum, TValue> :
         where TEnum : SmartEnum<TEnum, TValue>
         where TValue : IEquatable<TValue>, IComparable<TValue>
     {
+        static readonly Lazy<IReadOnlyCollection<TEnum>> _enumOptions = 
+            new Lazy<IReadOnlyCollection<TEnum>>(
+                SmartEnumOptions<TEnum, TValue>.GetAll,
+                LazyThreadSafetyMode.ExecutionAndPublication);
+        
         static readonly Lazy<Dictionary<string, TEnum>> _fromName = 
-            new Lazy<Dictionary<string, TEnum>>(() => GetAllOptions().ToDictionary(item => item.Name));
+            new Lazy<Dictionary<string, TEnum>>(() => _enumOptions.Value.ToDictionary(item => item.Name));
 
         static readonly Lazy<Dictionary<string, TEnum>> _fromNameIgnoreCase = 
-            new Lazy<Dictionary<string, TEnum>>(() => GetAllOptions().ToDictionary(item => item.Name, StringComparer.OrdinalIgnoreCase));
+            new Lazy<Dictionary<string, TEnum>>(() => _enumOptions.Value.ToDictionary(item => item.Name, StringComparer.OrdinalIgnoreCase));
 
         static readonly Lazy<Dictionary<TValue, TEnum>> _fromValue = 
             new Lazy<Dictionary<TValue, TEnum>>(() => {
                 // multiple enums with same value are allowed but store only one per value
                 var dictionary = new Dictionary<TValue, TEnum>();
-                foreach (var item in GetAllOptions())
+                foreach (var item in _enumOptions.Value)
                 {
                     if (!dictionary.ContainsKey(item._value))
                         dictionary.Add(item._value, item);
                 }
                 return dictionary;
             });
 
-        private static IEnumerable<TEnum> GetAllOptions()
-        {
-            Type baseType = typeof(TEnum);
-            IEnumerable<Type> enumTypes = Assembly.GetAssembly(baseType).GetTypes().Where(t => baseType.IsAssignableFrom(t));
-
-            List<TEnum> options = new List<TEnum>();
-            foreach (Type enumType in enumTypes)
-            {
-                List<TEnum> typeEnumOptions = enumType.GetFieldsOfType<TEnum>();
-                options.AddRange(typeEnumOptions);
-            }
-
-            return options.OrderBy(t => t.Name).ToList();
```

Comment:
This OrderBy thing in combination with #99 means that when multiple derived enums with same value exist only the first in alphabetical order will be resolved.